### PR TITLE
Enumeration and iteration for DFA

### PR DESCRIPTION
**Pull request by Tagl**
Code for #79 
I had to update the copy function in automaton to skip copying the caches. Don't really like the fix I made.
Rest is just as discussed and then added min and max word length functions which are calculated with BFS and longest path in DAG, respectively.

### CODE REVIEW

**Comment by caleb531** [comment]
@Tagl Thanks so much for this! Will take a closer look at the code tomorrow.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:595]: This entry in the cache is accessed a lot, and `i` is not modified in this loop. You should factor out some of these lookups, as they can get expensive.

**Review by eliotwrobson — COMMENTED** [review]
Looks good! A few notes on efficiency. Also, @caleb531 since some of the information is shared between functions here and existing ones (like generating the digraph and stuff), wonder if those should be made cached properties. Even for Python 3.7, we can use a back-port: https://pypi.org/project/cached-property/

**Comment by caleb531** [comment]
@eliotwrobson You know, it's funny you mention `@cached_property`, because before the whole immutability discussion, I was going to implement that for `NFA.lambda_closures`. But that was before I realized cached properties were only supported natively in Python 3.9 or newer. Fortunately, we don't need that anymore in the case of `lambda_closures` because NFAs are now immutable anyway.

But thinking about it more, I'm fine with adding the backport to this library so we can leverage cached properties in places where they're really needed. Especially since `@lru_cache` (which _is_ available in Python 3.7) is [not friendly to class instances](https://rednafi.github.io/reflections/dont-wrap-instance-methods-with-functoolslru_cache-decorator-in-python.html).

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Good point, updated

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Agreed. With immutability we will want to cache the results of some of these functions such as isempty, isfinite, maximum_word_length and minimum_word_length.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: I now only construct it in `maximum_word_length` and then `isfinite` calls `maximum_word_length`.

File: tests/test_dfa.py
Comment (eliotwrobson) [inline tests/test_dfa.py:1311]: Maybe not totally needed, but I merged a function that creates a minimal DFA accepting a finite language, so a cool test would be to feed in a finite language to that and check that your code here is able to recover it.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:582]: Maybe a bit extra, but I think you can do the inside of both of the loops with comprehensions.

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1311]: Agreed

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:561]:
```suggestion
                    for suffix_state in self.transitions[state].values():
```

Comment (Tagl) [reply automata/fa/dfa.py]: Fixed

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Like this? Or you had something else in mind?

File: automata/base/automaton.py
Comment (caleb531) [inline automata/base/automaton.py:70]: @Tagl Great idea to exclude the private-marked attributes.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:153]: @Tagl Could you please add docstrings to this and the other magic methods below?

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:159]: @Tagl I would much prefer a `for` loop here to eliminate the manual counting:

```python
def __iter__(self):
    min_len = self.minimum_word_length()
    max_len = self.maximum_word_length()
    for i in range(min_len, max_len + 1):
        yield from self.words_of_length(i)
```

Also, props to you for using `yield from`

**Comment by Tagl** [comment]
@caleb531 @eliotwrobson I'm thinking I'll limit the scope of this PR to just FA since those are both regular languages. I will therefore implement for NFA as well but make another PR later on for PDA.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:33]: @Tagl You should probably rebase on top of my latest commits to the `immutability` branch. I've made the DFA class immutable by overriding `__setattr__` and `__delattr__`—this looks to be using the old `DFA.__init__` code.

I would ask that you keep the `_word_cache` scoped to the methods that need it, passing it around as a function argument as needed.

**Review by caleb531 — CHANGES_REQUESTED** [review]
@Tagl A few minor things here and there, but I think the big request I have is related to refactoring `word_cache` to be scoped to the method rather than stored as an attribute on the entire instance.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:159]: The reason for the while loop is that `limit` can be `float('inf')`

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py:33]: @Tagl That's a fair point. And the fact that this is a private (and volatile, i.e. caching) property means the fact that it changes isn't such a big deal.

Then my recommendation would be: when you rebase off the latest `immutability` branch, feel free to create the cache initially using the `super().__init__` call, then because you should not need to reassign to `_word_cache`, you should be able to push to it freely.

This will probably break my existing `test_dfa_immutable_dict` test because in that test, I am assuming that all attrs on the DFA's `__dict__` are hashable. But I can probably adjust that, or maybe expose a separate method for retrieving the `__dict__` with internal properties excluded.

@eliotwrobson What do you think?

Comment (eliotwrobson) [reply automata/fa/dfa.py:33]: @caleb531 I think that course of action sounds reasonable. @Tagl Is right that these really should be stored along with the lifetime of the object, since having to recompute them is expensive, and it really should be done as-needed (the way that it is done here). I think testing for immutability excluding internal properties is a good plan.

**Comment by caleb531** [comment]
@Tagl If it's alright, I'm going to rebase your `enumeration` branch on top of the latest `immutability` branch.

**Comment by caleb531** [comment]
@Tagl Please don't push any more commits to `enumeration` for the time being. I am still rebasing.

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py]: @Tagl I see. `range()` can't receive `float` values, so I understand the need to use a `while` now.

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py]: @eliotwrobson Agreed. I've also added a new `Automaton.attributes()` method to this commit that basically returns `__dict__` but with only the public attributes. I might rename this later to be more semantically-clear, but not until after this PR is merged.

**Comment by caleb531** [comment]
@Tagl Okay, rebase is complete. I'm all done working on the `enumeration` branch.

The last thing I would really like to see in this PR is documentation for these new public methods in the README. I realized we'll all need to make sure to keep the README up to date with all these PRs being merged, otherwise it'll become a major burden on me to put that all together. I would ask if we could update the README incrementally (as needed) with each PR as it has been in the past.

@eliotwrobson would you have time to send me a separate PR with README updates for your recently-added public methods? (any README updates for #82 can just be part of that PR). That would be a great help to me.

**Comment by eliotwrobson** [comment]
@caleb531 Sure, I'll look through the last few PRs I've added and update the README on #82. There's a slight change on the public methods for #77 (adding a couple of parameters), so I'll make the update for that one localized there.

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py]: I added a test for iter that does this

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:721]: Can you have this throw an exception instead? I want to write a class method that returns the empty DFA later.

Similarly, I think this algorithm doesn't work with the empty string, so it should throw an exception in that case.

Comment (Tagl) [reply automata/fa/dfa.py:721]: One of my tests for len was using the function to create the empty language, which is finite, and therefore should be possible to create with this function in my opinion. I don't think we should be raising an exception in that case.

The algorithm does work with the empty string, at least what I've tried so far seems to work.

Comment (Tagl) [reply automata/fa/dfa.py:721]: However I do agree with adding a specific function for creating the empty language and this function can then defer to that function for this edge case.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:717]: I think this is here by accident? We removed this function a little while ago.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:683]:
```suggestion
        while queue:
```

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:729]: Sounds good. Did you try the empty string with at least one other string? I think that might be the situation that causes problems.

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Merge accident

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Updated

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:724]: I tried the language with only the empty word and then two other finite languages that were larger but also included the empty word.

**Comment by Tagl** [comment]
@caleb531 Since this contains some common bugfixes too I'll leave the PR as just DFA and make another one for NFA so you can get the bugfixes merged sooner.

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py]: @Tagl Oh, I see. Sorry about that—that was an artifact I must've introduced during the rebase.

**Review by eliotwrobson — APPROVED** [review]
Looks good! I think it'll be good to merge this then integrate it with the new methods in the DFA construction.

**Comment by Tagl** [comment]
@eliotwrobson I just made a pretty important change to the generation. I was under the impression that dict and set iteration followed insertion order since Python 3.6 but I was getting seemingly random orders. Since each value is guaranteed to only generate once we can use lists instead for the cache and then maintain proper lexicographical ordering.

**Comment by eliotwrobson** [comment]
@Tagl sounds good. If you think using a list will affect performance, I think you could use an ordered dict instead.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:651]: @Tagl (cc @eliotwrobson) If you updated the `defaultdict` to use a `list` instead of a `set`, shouldn't this dict comprehension also follow suit?

```python
level.update({state: [''] for state in self.final_states})
```

File: tests/test_dfa.py
Comment (caleb531) [inline tests/test_dfa.py:1490]: @Tagl Couldn't this be rewritten as:

```python
generated_list = list(A)
```

I think that would be more consistent with how the user will actually consume the iterable.

File: tests/test_dfa.py
Comment (caleb531) [inline tests/test_dfa.py:1472]: @Tagl Could you please rename `A` to something like `dfa` or `dfa1`? The capital letter makes me think `A` is a class rather than an instance.

File: automata/base/automaton.py
Comment (caleb531) [inline automata/base/automaton.py:118]: @Tagl Since `other` does not represent another automaton instance, but rather, the input string, could you rename `other` to `input_str` here? This would be more consistent with methods like `read_input_stepwise`.

**Review by caleb531 — CHANGES_REQUESTED** [review]
@Tagl This is looking great overall, just have a few style requests (and one question related to the `defaultdict`).

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py:651]: Yes

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1490]: No, it can't be done here because the iteration is infinite so I only iterate as many times as there are items in `expected`.

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1472]: Will do

File: automata/fa/dfa.py
Comment (Tagl) [reply automata/fa/dfa.py]: Updated

File: tests/test_dfa.py
Comment (Tagl) [reply tests/test_dfa.py:1472]: Updated

File: automata/base/automaton.py
Comment (Tagl) [reply automata/base/automaton.py]: Updated

File: tests/test_dfa.py
Comment (caleb531) [reply tests/test_dfa.py:1490]: @Tagl Ah yes, I'm not used to working with potentially-infinite generators 😅. Makes sense.

File: tests/test_dfa.py
Comment (caleb531) [reply tests/test_dfa.py]: Thanks, although I think there's still one hanging around:

Screenshot 2022-10-27 at 1 42 56 PM

Comment (Tagl) [reply tests/test_dfa.py]: Missed one originally, updated again.

**Review by caleb531 — APPROVED** [review]
@Tagl This all looks great now! Thank you so much for writing this and for making my requested changes.

**Pull request merged**
